crates/bitwarden-crypto/src/safe/password_protected_key_envelope.rs

```diff
@@ -0,0 +1,572 @@
+use std::{marker::PhantomData, num::TryFromIntError};
+
+use argon2::Params;
+use ciborium::{value::Integer, Value};
+use coset::{
+    iana::CoapContentFormat, CborSerializable, ContentType, CoseError, Header, HeaderBuilder,
+};
+use rand::RngCore;
+use thiserror::Error;
+
+use crate::{
+    cose::{
+        extract_bytes, extract_integer, CoseExtractError, ALG_ARGON2ID13, ARGON2_ITERATIONS,
+        ARGON2_MEMORY, ARGON2_PARALLELISM, ARGON2_SALT, CONTENT_TYPE_BITWARDEN_LEGACY_KEY,
+    },
+    xchacha20, BitwardenLegacyKeyBytes, ContentFormat, CoseKeyBytes, EncodedSymmetricKey, KeyIds,
+    KeyStoreContext, SymmetricCryptoKey,
+};
+
+/// A password-protected key envelope can seal a symmetric key, and protect it with a password. It
+/// does so by using a Key Derivation Function (KDF), to increase the difficulty of brute-forcing
+/// the password.
+///
+/// The KDF parameters such as iterations and salt are stored in the key-envelope and do not have to
+/// be provided.
+///
+/// Internally, Argon2 as the KDF, and XChaCha20-Poly1305 are used to encrypt the message.
+pub struct PasswordProtectedKeyEnvelope<Ids: KeyIds> {
+    _phantom: PhantomData<Ids>,
+    cose_encrypt: coset::CoseEncrypt,
+}
+
+impl<Ids: KeyIds> PasswordProtectedKeyEnvelope<Ids> {
+    /// Seals a symmetric key with a password, using the current default KDF parameters and a random
+    /// salt.
+    ///
+    /// This should never fail, except for memory allocation error, when running the KDF.
+    pub fn seal(
+        key_to_seal: Ids::Symmetric,
```

Comment:
```suggestion
        keyslot_to_seal: Ids::Symmetric,
```
Something to make it clearer, this is the ID of the key. Maybe this is fine how it is and we just rely on the type system.

Reply:
`key` alone should generally not be used, sorry my phrasing above was a bit confusing because it ended up suggesting that. I meant `key` + purpose, or `key_id` + purpose, so something like `wrapping_key`, or `key_to_be_wrapped`, but *never* key alone.


On the point of function arguments, in rust function arguments are generally not documented, and they (rust team) have been explicit about removing an examples around it: https://bitwarden.slack.com/archives/C054ZQSBS49/p1754994308725229

The referenced code was added by me, while I still did not know about this. The contributing docs are still lacking on this end currently.